community/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/compatibility/v3_3/SemanticTableConverter.scala

```diff
@@ -0,0 +1,99 @@
+      astV3_4.ASTAnnotationMap.empty,
+      new mutable.HashMap[String, utilV3_4.LabelId],
+      new mutable.HashMap[String, utilV3_4.PropertyKeyId],
+      convert(table.resolvedRelTypeNames))
+  }
+
+
+  private def convert(types: astV3_3.ASTAnnotationMap[astV3_3.Expression, frontendV3_3.ExpressionTypeInfo], expressionMapping: ExpressionMapping3To4):
+  astV3_4.ASTAnnotationMap[expressionsV3_4.Expression, frontendV3_4.semantics.ExpressionTypeInfo] = {
+    val result: Map[ExpressionV3_4, ExpressionTypeInfo] = types.filter {
+      case (exprV3_3, _) => expressionMapping.isDefinedAt((exprV3_3, exprV3_3.position))
+    }.map {
+      case (exprV3_3, typeInfoV3_3) => (expressionMapping((exprV3_3, exprV3_3.position)), convert(typeInfoV3_3))
+    }
+    astV3_4.ASTAnnotationMap(result.toSeq:_*)
+  }
+
+  private def convert(resolvedRelTypeNames: mutable.Map[String, frontendV3_3.RelTypeId]): mutable.Map[String, utilV3_4.RelTypeId] = {
+    val res = new mutable.HashMap[String, utilV3_4.RelTypeId]
+    resolvedRelTypeNames.foreach {
+      case (key, r3_3) => res += ((key, utilV3_4.RelTypeId(r3_3.id)))
+    }
+    res
+  }
+
+
+  private def convert(typeInfoV3_3: frontendV3_3.ExpressionTypeInfo): frontendV3_4.semantics.ExpressionTypeInfo = typeInfoV3_3 match {
+    case frontendV3_3.ExpressionTypeInfo(specified, expected) =>
+      frontendV3_4.semantics.ExpressionTypeInfo(convert(specified), expected.map(convert _))
+  }
+
+  private def convert(specified: frontendV3_3.symbols.TypeSpec): utilV3_4.symbols.TypeSpec = {
+      new utilV3_4.symbols.TypeSpec(specified.ranges.map(convert _))
+  }
+
+  private def convert(range: frontendV3_3.symbols.TypeRange): utilV3_4.symbols.TypeRange = range match {
+    case frontendV3_3.symbols.TypeRange(lower, upper) =>
+      utilV3_4.symbols.TypeRange(convert(lower), upper.map(convert _))
+  }
+
+  private def convert(cypherType: frontendV3_3.symbols.CypherType): utilV3_4.symbols.CypherType = cypherType match {
+    case frontendV3_3.symbols.CTAny => utilV3_4.symbols.CTAny
+    case frontendV3_3.symbols.CTBoolean => utilV3_4.symbols.CTBoolean
+    case frontendV3_3.symbols.CTFloat => utilV3_4.symbols.CTFloat
+    case frontendV3_3.symbols.CTGeometry => utilV3_4.symbols.CTGeometry
+    case frontendV3_3.symbols.CTGraphRef => utilV3_4.symbols.CTGraphRef
+    case frontendV3_3.symbols.CTInteger => utilV3_4.symbols.CTInteger
+    case frontendV3_3.symbols.ListType(iteratedType) => utilV3_4.symbols.CTList(convert(iteratedType))
+    case frontendV3_3.symbols.CTMap => utilV3_4.symbols.CTMap
+    case frontendV3_3.symbols.CTNode => utilV3_4.symbols.CTNode
+    case frontendV3_3.symbols.CTNumber => utilV3_4.symbols.CTNumber
+    case frontendV3_3.symbols.CTPath => utilV3_4.symbols.CTPath
+    case frontendV3_3.symbols.CTPoint => utilV3_4.symbols.CTPoint
+    case frontendV3_3.symbols.CTRelationship => utilV3_4.symbols.CTRelationship
+    case frontendV3_3.symbols.CTString => utilV3_4.symbols.CTString
+  }
```

Comment:
This could possibly be share implementation with the type conversion in the `LogicalPlanConverter`, if that's reasonably smooth.

Reply:
We decided against that, given that here the number of cases is reasonably small, so that we do not want to pay the reflection overhead.

Reply:
I was thinking to call the other way, but no problem - do as you want =)